File: e2e/servers/python/mcp/main.py
"""MCP E2E Test Server with x402 Payment-Wrapped Tools.

Thin MCP adapter over the same mechanisms catalog the HTTP frameworks use: one
tool per resolved route, each wrapped with ``create_payment_wrapper`` using
payment requirements built from the same ``accepts`` config
``build_payment_routes`` feeds the HTTP middleware. Tools take no arguments
and return the fixed ``{message, timestamp}`` body every HTTP route returns.
"""

from __future__ import annotations

import os
import threading
from typing import Any


def main() -> None:
    """Start the MCP server with x402 payment-wrapped tools."""
    import uvicorn
    from mcp.server.fastmcp import FastMCP
    from starlette.applications import Starlette
    from starlette.responses import JSONResponse
    from starlette.routing import Route

    from x402 import ResourceConfig, ResourceInfo, x402ResourceServer
    from x402.http import FacilitatorConfig, HTTPFacilitatorClient
    from x402.mcp import create_payment_wrapper

    from catalog import PROTECTED_ROUTE_MESSAGE, catalog_routes, mcp_tool_name, resolve_routes, route_description
    from config import build_resolved_route_config, configure_resource_server, load_server_config
    from handlers import CLOSE_PATH, HEALTH_PATH, close_body, health_body, route_body

    cfg = load_server_config()

    mcp = FastMCP("x402 MCP E2E Server")

    facilitator_client = HTTPFacilitatorClient(FacilitatorConfig(url=cfg.facilitator_url))
    resource_server = x402ResourceServer(facilitator_client)
    configure_resource_server(resource_server, cfg)
    resource_server.initialize()

    # Tool descriptions come from the unresolved catalog routes (which still
    # carry `asset_transfer_method`), so unconfigured networks get sensible
    # descriptions too even though they never register a tool below.
    tool_descriptions = {
        route.path: route_description(
            route.network, route.scheme, route.asset_transfer_method, route.extensions, route.payment_flow
        )
        for route in catalog_routes()
    }

    def register_route_tool(route: Any) -> None:
        tool_name = mcp_tool_name(route.path)
        description = tool_descriptions.get(route.path, f"Paid MCP tool for {route.path}")
        route_config = build_resolved_route_config(route, "mcp")
        accepts_cfg = route_config["accepts"]
        accepts = resource_server.build_payment_requirements(
            ResourceConfig(
                scheme=accepts_cfg["scheme"],
                pay_to=accepts_cfg["payTo"],
                network=accepts_cfg["network"],
                price=accepts_cfg["price"],
                extra=accepts_cfg.get("extra"),
            )
        )
        wrapper = create_payment_wrapper(
            resource_server,
            accepts=accepts,
            resource=ResourceInfo(
                url=f"mcp://tool/{tool_name}",
                description=description,
                mime_type="application/json",
            ),
            extensions=route_config.get("extensions"),
        )

        @mcp.tool(name=tool_name, description=description)
        @wrapper
        async def _tool() -> str:
            import json
            from datetime import datetime, timezone

            return json.dumps(
                {
                    "message": PROTECTED_ROUTE_MESSAGE,
                    "timestamp": datetime.now(timezone.utc).isoformat(),
                }
            )

    for route in resolve_routes():
        register_route_tool(route)

    @mcp.tool(name="ping", description="A free health check tool")
    def ping() -> str:
        return "pong"

    async def health(request):
        return JSONResponse(health_body("mcp"))

    async def close(request):
        response = JSONResponse(close_body())

        def shutdown():
            import time

            time.sleep(0.1)
            os._exit(0)

        threading.Thread(target=shutdown, daemon=True).start()
        return response

    async def on_startup() -> None:
        # Emitted from the ASGI startup hook (not before uvicorn.run()) so the
        # "Server listening" log only appears once the socket is actually
        # bound and the app is ready to accept requests.
        print(f"Server listening on port {cfg.port}", flush=True)
        print(f"SSE endpoint: http://localhost:{cfg.port}/sse", flush=True)
        print(f"Health: http://localhost:{cfg.port}{HEALTH_PATH}", flush=True)

    mcp_app = mcp.sse_app()

    app = Starlette(
        routes=[
            Route(HEALTH_PATH, health, methods=["GET"]),
            Route(CLOSE_PATH, close, methods=["POST"]),
        ],
        on_startup=[on_startup],
    )

    # Mount MCP SSE app at root so /sse and /messages work
    app.mount("/", mcp_app)

    uvicorn.run(app, host="0.0.0.0", port=cfg.port, log_level="warning")


if __name__ == "__main__":
    main()

File: e2e/servers/go/mcp/main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/google/jsonschema-go/jsonschema"
	"github.com/modelcontextprotocol/go-sdk/mcp"

	e2eserver "github.com/x402-foundation/x402/e2e/servers/go"
	x402 "github.com/x402-foundation/x402/go/v2"
	mcp402 "github.com/x402-foundation/x402/go/v2/mcp"
)

var shutdownRequested bool

// MCP E2E Test Server with x402 Payment Support.
//
// Thin MCP transport over the same mechanisms catalog as the HTTP servers
// (see e2eserver.ResolvedRoutes) — each catalog route becomes an MCP tool
// wrapped with mcp402.PaymentWrapper for verification + settlement.

func main() {
	cfg := e2eserver.LoadConfig()
	facilitatorClient := e2eserver.NewFacilitatorClient(cfg)

	resourceServer := x402.Newx402ResourceServer(x402.WithFacilitatorClient(facilitatorClient))
	for _, binding := range e2eserver.SchemeBindings(cfg) {
		resourceServer.Register(binding.Network, binding.Server)
	}

	ctx := context.Background()
	if err := resourceServer.Initialize(ctx); err != nil {
		fmt.Printf("Warning: failed to initialize x402 server: %v\n", err)
	}

	mcpServer := mcp.NewServer(&mcp.Implementation{
		Name:    "x402 MCP E2E Server",
		Version: "1.0.0",
	}, nil)

	for _, route := range e2eserver.ResolvedRoutes() {
		route := route
		toolName := e2eserver.McpToolName(route.Path)
		description := e2eserver.RouteDescription(route)
		_, extensions := e2eserver.BuildResolvedRouteConfig(route, "mcp")

		requirements, err := resourceServer.BuildPaymentRequirementsFromConfig(ctx, x402.ResourceConfig{
			Scheme:  route.Scheme,
			PayTo:   route.PayTo,
			Price:   route.Price,
			Network: x402.Network(route.Network),
			Extra:   route.Extra,
		})
		if err != nil {
			fmt.Printf("❌ Failed to build payment requirements for %s: %v\n", route.Path, err)
			os.Exit(1)
		}

		wrapperConfig := mcp402.PaymentWrapperConfig{
			Accepts: requirements,
			Resource: &mcp402.ResourceInfo{
				URL:         "mcp://tool/" + toolName,
				Description: description,
				MimeType:    "application/json",
			},
		}
		if len(extensions) > 0 {
			wrapperConfig.Extensions = extensions
		}
		wrapper := mcp402.NewPaymentWrapper(resourceServer, wrapperConfig)

		tool := &mcp.Tool{
			Name:        toolName,
			Description: description,
			InputSchema: &jsonschema.Schema{Type: "object"},
		}

		mcpServer.AddTool(tool, wrapper.Wrap(
			func(ctx context.Context, request *mcp.CallToolRequest) (*mcp.CallToolResult, error) {
				if shutdownRequested {
					body, _ := json.Marshal(map[string]interface{}{"error": "Server shutting down"})
					return &mcp.CallToolResult{
						Content: []mcp.Content{&mcp.TextContent{Text: string(body)}},
						IsError: true,
					}, nil
				}
				body, _ := json.Marshal(e2eserver.RouteBody())
				return &mcp.CallToolResult{Content: []mcp.Content{&mcp.TextContent{Text: string(body)}}}, nil
			},
		))
	}

	mux := http.NewServeMux()

	mux.HandleFunc("GET /health", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, http.StatusOK, e2eserver.HealthBody())
	})

	mux.HandleFunc("POST /close", func(w http.ResponseWriter, r *http.Request) {
		shutdownRequested = true

		writeJSON(w, http.StatusOK, map[string]interface{}{
			"message": "Server shutting down gracefully",
		})
		fmt.Println("Received shutdown request")

		go func() {
			time.Sleep(100 * time.Millisecond)
			os.Exit(0)
		}()
	})

	sseHandler := mcp.NewSSEHandler(func(r *http.Request) *mcp.Server {
		return mcpServer
	}, nil)
	mux.Handle("/", sseHandler)

	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-quit
		fmt.Println("Received shutdown signal, exiting...")
		os.Exit(0)
	}()

	// Bind the socket first and only then print the "listening" log, so the
	// e2e harness (which treats this log line as the readiness signal)
	// doesn't consider the server ready before it can actually accept
	// connections.
	listener, err := net.Listen("tcp", ":"+cfg.Port)
	if err != nil {
		fmt.Printf("Error starting server: %v\n", err)
		os.Exit(1)
	}

	fmt.Println(e2eserver.FormatStartupBanner(
		"x402 MCP E2E Test Server",
		"http://localhost:"+cfg.Port,
	))

	if err := http.Serve(listener, mux); err != nil {
		fmt.Printf("Error starting server: %v\n", err)
		os.Exit(1)
	}
}

// writeJSON is a helper to write JSON responses.
func writeJSON(w http.ResponseWriter, status int, data interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(data)
}
